---
title: "Bash-Projects_NGS"
author: "Justina Kraujūnaitė"
date: "3/26/2021"
output:
    md_document:
        variant: markdown_github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(engine.opts = list(bash = "-l"))
```


# Project 1

## Basic Unix commands

Assume you sequenced and assembled the genome of Malus domestica (apple), and performed gene annotation. You then collected samples and ran RNA-seq experiments to determine sets of genes that are expressed in the various tissues. This information was stored, respectively, in the following files: “apple.genome”, “apple.genes”, “apple.condition{A,B,C}”.

NOTE: The apple genome and the apple gene annotations for this project were extracted from the Rosaceae Genome Database (RGD). Actual data have then been modified, and hence may not directly reflect the information in the original RGD records.

Data extraction from the tar archive:
```{bash}
tar -xvf gencommand_proj1_data.tar
```

**1. How many chromosomes are there in the genome?**
```{bash}
grep -c ">" apple.genome
```

**2. How many genes?**
```{bash}
cut -f1 apple.genes | sort -u | wc -l
```

**3. How many transcript variants?**
```{bash}
cut -f2 apple.genes | sort -u | wc -l
```

**4. How many genes have a single splice variant?**
```{bash}
cut -f1 apple.genes | uniq -c | grep " 1 " | wc -l
```

**5. How many genes have 2 or more splice variants?**
```{bash}
cut -f1 apple.genes | uniq -c | grep -v " 1 " | wc -l
```

**6. How many genes are there on the ‘+’ strand?**
```{bash}
cut -f1,4 apple.genes | sort | uniq -c | grep "+" | wc -l
```

**7. How many genes are there on the ‘-’ strand?**
```{bash}
cut -f1,4 apple.genes | sort | uniq -c | grep "-" | wc -l
```

**8. How many genes are there on chromosome chr1?**  
**9. How many genes are there on each chromosome chr2?**  
**10. How many genes are there on each chromosome chr3?**
```{bash}
cut -f1,3 apple.genes | sort -u | cut -f2 | sort | uniq -c
```

**11. How many transcripts are there on chr1?**  
**12. How many transcripts are there on chr2?**  
**13. How many transcripts are there on chr3?**
```{bash}
cut -f2,3 apple.genes | sort -u | cut -f2 | sort | uniq -c
```

**14. How many genes are in common between condition A and condition B?**
```{bash}
cut -f1 apple.conditionA | sort -u > apple.conditionA.sorted
cut -f1 apple.conditionB | sort -u > apple.conditionB.sorted
comm -1 -2 apple.conditionA.sorted apple.conditionB.sorted | wc -l
```

**15. How many genes are specific to condition A?**
```{bash}
comm -2 -3 apple.conditionA.sorted apple.conditionB.sorted | wc -l
```

**16. How many genes are specific to condition B?**
```{bash}
comm -1 -3 apple.conditionA.sorted apple.conditionB.sorted | wc -l
```

**17. How many genes are in common to all three conditions?**
```{bash}
cut -f1 apple.conditionC | sort -u > apple.conditionC.sorted
comm -1 -2 apple.conditionA.sorted apple.conditionB.sorted > conditionsAB
comm -1 -2 conditionsAB apple.conditionC.sorted | wc -l
```

# Project 2

## SAMtools and BEDtools

As part of a larger project cataloging genetic variation in the plant Arabidopsis thaliana, you sequenced and assembled the genome of one strain (‘wu_0_A’), then mapped back the reads to the assembled genome. The resulting BAM file is included in the package ‘gencommand_proj2_data.tar.gz’.

NOTE: Input data have been obtained and modified from those generated by the 1001 Genomes Project, accession ‘Wu_0_A’.

Apply these rules and steps to the questions marked above each rule. 

Questions 1 - 5:  
for the original set of alignments (file ‘athal_wu_0_A.bam’).

Questions 6 - 10:  
extract only the alignments in the range “Chr3:11,777,000-11,794,000”, corresponding to a locus of interest.

Questions 11 - 15:  
determine general information about the alignment process from the original BAM file. 

Questions 16 - 20:  
using BEDtools, examine how many of the alignments at point 2 overlap exons at the locus of interest. Use the BEDtools ‘-wo’ option to only report non-zero overlaps. The list of exons is given in the included ‘athal_wu_0_A_annot.gtf’ GTF file.

Data extraction from the tar archive:
```{bash}
tar -xvf gencommand_proj2_data.tar
```

**1. How many alignments does the set contain?**
```{bash}
samtools view athal_wu_0_A.bam | wc -l
```

**2. How many alignments show the read’s mate unmapped?**
```{bash}
samtools view athal_wu_0_A.bam | cut -f7 | grep "*" | wc -l
```

**3. How many alignments contain a deletion (D)?**
```{bash}
samtools view athal_wu_0_A.bam | cut -f6 | grep "D" | wc -l
```

**4. How many alignments show the read’s mate mapped to the same chromosome?**
```{bash}
samtools view athal_wu_0_A.bam | cut -f7 | grep "=" | wc -l
```

**5. How many alignments are spliced?**
```{bash}
samtools view athal_wu_0_A.bam | cut -f6 | grep "N" | wc -l
```

**6. How many alignments does the set contain?**
```{bash}
# sorting the .bam file
samtools sort athal_wu_0_A.bam -o athal_wu_0_A.sorted.bam
# indexing sorted .bam file
samtools index athal_wu_0_A.sorted.bam
# extracting th range of interest
samtools view athal_wu_0_A.sorted.bam "Chr3:11,777,000-11,794,000" | wc -l
```

**7. How many alignments show the read’s mate unmapped?**
```{bash}
samtools view athal_wu_0_A.sorted.bam "Chr3:11,777,000-11,794,000" | cut -f7 | grep "*" | wc -l
```

**8. How many alignments contain a deletion (D)?**
```{bash}
samtools view athal_wu_0_A.sorted.bam "Chr3:11,777,000-11,794,000" | cut -f6 | grep "D" | wc -l
```

**9. How many alignments show the read’s mate mapped to the same chromosome?**
```{bash}
samtools view athal_wu_0_A.sorted.bam "Chr3:11,777,000-11,794,000" | cut -f7 | grep "=" | wc -l
```

**10. How many alignments are spliced?**
```{bash}
samtools view athal_wu_0_A.sorted.bam "Chr3:11,777,000-11,794,000" | cut -f6 | grep "N" | wc -l
```

**11. How many sequences are in the genome file?** (Answer: 7)  
**12. What is the length of the first sequence in the genome file?** (Answer: 29923332)  
**13. What alignment tool was used?** (Answer: stampy)
```{bash}
samtools view -H athal_wu_0_A.bam
```

**14. What is the read identifier (name) for the first alignment?** (Answer: GAII05_0002:1:113:7822:3886#0)   
**15. What is the start position of this read’s mate on the genome? Give this as ‘chrom:pos’ if the read was mapped, or ‘\*\’ if unmapped.** (Answer: Chr3:11700332)
```{bash}
samtools view athal_wu_0_A.bam | head -n 1
```

**16. How many overlaps (each overlap is reported on one line) are reported?**
```{bash}
bedtools bamtobed -i athal_wu_0_A.bam > athal_wu_0_A.bed
bedtools intersect -wo -a athal_wu_0_A_annot.gtf -b athal_wu_0_A.bed | wc -l
```

**17. How many of these are 10 bases or longer?**
```{bash}
bedtools intersect -wo -a athal_wu_0_A_annot.gtf -b athal_wu_0_A.bed | cut -f16 | awk '$1 > 9' | wc -l
```

**18. How many alignments overlap the annotations?**
```{bash}
bedtools intersect -wo -a athal_wu_0_A_annot.gtf -b athal_wu_0_A.bed | wc -l
```

**19. Conversely, how many exons have reads mapped to them?**
```{bash}
bedtools intersect -wo -a athal_wu_0_A_annot.gtf -b athal_wu_0_A.bed | cut -f4 | sort -u | wc -l
```

**20. If you were to convert the transcript annotations in the file “athal_wu_0_A_annot.gtf” into BED format, how many BED records would be generated?**
```{bash}
bedtools intersect -wo -a athal_wu_0_A_annot.gtf -b athal_wu_0_A.bed | cut -f9 | cut -d " " -f4 | sort -u | wc -l
```

# Project 3

## Bowtie2 and BCFtools

As part of the effort to catalog genetic variation in the plant Arabidopsis thaliana, you re-sequenced the genome of one strain (‘wu_0_A’; genome file: ‘wu_0.v7.fas’), to determine genetic variants in this organism. The sequencing reads produced are in the file ‘wu_0_A_wgs.fastq’. Using the tools bowtie2, samtools and bcftools, develop a pipeline for variant calling in this genome. 

NOTE: Genome and re-sequencing data have been obtained and modified from those generated by the 1001 Genomes Project, accession ‘Wu_0_A’.

Apply these rules and steps to the questions marked above each rule.

Questions 1 - 5:  
generate a bowtie2 index of the wu_0_A genome using bowtie2-build, with the prefix 'wu_0'. 

Questions 6 - 14:  
run bowtie2 to align the reads to the genome, under two scenarios: first, to report only full-length matches of the reads; and second, to allow partial (local) matches. All other parameters are as set by default. 

For the following set of questions (11 - 20), use the set of full-length alignments calculated under scenario 1 only. Convert this SAM file to BAM, then sort the resulting BAM file. 

Questions 15 - 19:  
compile candidate sites of variation using SAMtools mpileup for further evaluation with BCFtools. Provide the reference fasta genome and use the option “-uv” to generate the output in uncompressed VCF format for easy examination. 

Questions 20 - 24:  
call variants using ‘BCFtools call’ with the multiallelic-caller model. For this, you will need to first re-run SAMtools mpileup with the BCF output format option (‘-g’) to generate the set of candidate sites to be evaluated by BCFtools. In the output to BCFtools, select to show only the variant sites, in uncompressed VCF format for easy examination.

Data extraction from the tar archive:
```{bash}
tar -xvf gencommand_proj3_data.tar
```

**1. How many sequences were in the genome?**
```{bash}
cat wu_0.v7.fas | grep ">" | wc -l
```

**2. What was the name of the third sequence in the genome file? Give the name only, without the “>” sign.** (Answer: Chr3)  
**3. What was the name of the last sequence in the genome file? Give the name only, without the “>” sign.** (Answer: mitochondria)
```{bash}
cat wu_0.v7.fas | grep ">"
```

**4. How many index files did the operation create?** (Answer: 6)  
**5. What is the 3-character extension for the index files created?** (Answer: bt2)
```{bash results="hide"}
mkdir wu_0
bowtie2-build wu_0.v7.fas wu_0/wu_0
```
```{bash}
ls wu_0
```

**6. How many reads were in the original fastq file?**
```{bash}
cat wu_0_A_wgs.fastq | wc -l
```

**7. How many matches (alignments) were reported for the original (full-match) setting? Exclude lines in the file containing unmapped reads.** (Answer: 137719)  
**9. How many reads were mapped in the scenario in Question 7?** (Answer: 137719)  
**11. How many reads had multiple matches in the scenario in Question 7? You can find this in the bowtie2 summary; note that by default bowtie2 only reports the best match for each read.** (Answer: 43939)
```{bash}
bowtie2 -x wu_0/wu_0 wu_0_A_wgs.fastq -S wu_0.bt2.sam
```

**8. How many matches (alignments) were reported with the local-match setting? Exclude lines in the file containing unmapped reads.** (Answer: 141044)  
**10. How many reads were mapped in the scenario in Question 8?** (Answer: 141044)  
**12. How many reads had multiple matches in the scenario in Question 8? Use the format above. You can find this in the bowtie2 summary; note that by default bowtie2 only reports the best match for each read.** (Answer: 56105)
```{bash}
bowtie2 --local -x wu_0/wu_0 wu_0_A_wgs.fastq -S wu_0.bt2.local.sam
```

**13. How many alignments contained insertions and/or deletions, in the scenario in Question 7?**
```{bash}
cat wu_0.bt2.sam | cut -f6 | grep "I" | grep "D" | wc -l
```

**14. How many alignments contained insertions and/or deletions, in the scenario in Question 8?**
```{bash}
cat wu_0.bt2.local.sam | cut -f6 | grep "I" | grep "D" | wc -l
```

**15. How many entries were reported for Chr3?**
```{bash results="hide"}
samtools view -bT wu_0.v7.fas wu_0.bt2.sam > wu_0.bt2.bam
samtools sort wu_0.bt2.bam -o wu_0.bt2.sorted.bam
samtools index wu_0.bt2.sorted.bam
samtools mpileup -uv -f wu_0.v7.fas wu_0.bt2.sorted.bam > wu_0.vcf
```
```{bash}
cat wu_0.vcf | cut -f1 | grep "Chr3" | wc -l
```

**16. How many entries have ‘A’ as the corresponding genome letter?**
```{bash}
cat wu_0.vcf | cut -f4 | awk '$1 == "A"' | wc -l
```

**17. How many entries have exactly 20 supporting reads (read depth)?**
```{bash}
cat wu_0.vcf | cut -f8 | grep "DP=20" | wc -l
```

**18. How many entries represent indels?**
```{bash}
cat wu_0.vcf | cut -f8 | grep "INDEL" | wc -l
```

**19. How many entries are reported for position 175672 on Chr1?**
```{bash}
cat wu_0.vcf | cut -f1,2 | grep "Chr1" | awk '$2 == "175672"' | wc -l
```

**20. How many variants are called on Chr3?**
```{bash results="hide"}
samtools mpileup -g -f wu_0.v7.fas wu_0.bt2.sorted.bam > wu_0.bcf
bcftools call -v -m -Oz -o wu_0.vcf.gz wu_0.bcf
```
```{bash}
zcat < wu_0.vcf.gz | cut -f1 | grep "Chr3" | wc -l
```

**21. How many variants represent an A->T SNP? If useful, you can use ‘grep –P’ to allow tabular spaces in the search term.**
```{bash}
zcat < wu_0.vcf.gz | cut -f4,5 | awk '$1 == "A" && $2 == "T"' | wc -l
```

**22. How many entries are indels?**
```{bash}
zcat < wu_0.vcf.gz | cut -f8 | grep "INDEL" | wc -l
```

**23. How many entries have precisely 20 supporting reads (read depth)?**
```{bash}
zcat < wu_0.vcf.gz | cut -f8 | grep "DP=20" | wc -l
```

**24. What type of variant (i.e., SNP or INDEL) is called at position 11937923 on Chr3?** (Answer: SNP)
```{bash}
zcat < wu_0.vcf.gz | grep "Chr3" | awk '$2 == "11937923"'
```
